darts/dataprocessing/transformers/boxcox.py

```diff
@@ -0,0 +1,102 @@
+    def __init__(self,
+                 name: str = "BoxCox",
+                 validators: Optional[Sequence[Validator]] = None):
+        """
+        Box-Cox data transformer.
+        See https://otexts.com/fpp2/transformations.html#mathematical-transformations for more information
+
+        Parameters
+        ----------
+        name
+            A specific name for the transformer
+        validators
+            Sequence of validators that will be called before transform()
+        """
+        super().__init__(name=name, validators=validators)
+        self._lmbda = None
+
+    def fit(self,
+            data: TimeSeries,
+            lmbda: Optional[Union[float, Sequence[float]]] = None,
+            optim_method='mle') -> 'BoxCox':
+        """
+        Sets the `lmbda` parameter value.
+
+        Parameters
+        ----------
+        data
+            The time series to fit on
+        lmbda
+            If None given, will automatically find an optimal value of lmbda (for each dimension
+            of the time series) using `scipy.stats.boxcox_normmax` with `method=optim_method`
+            If a single float is given, the same lmbda value will be used for all dimensions of the series.
+            Also allows to specify a different lmbda value for each dimension of the time series by passing
+            a sequence of values.
+
+        Returns
+        -------
+            Fitted transformer (self)
+        """
+        super().fit(data)
+
+        raise_if(not isinstance(optim_method, str) or optim_method not in ['mle', 'pearsonr'],
+                 "optim_method parameter must be either 'mle' or 'pearsonr'",
+                 logger)
+
+        if lmbda is None:
+            # Compute optimal lmbda for each dimension of the time series
+            lmbda = data.pd_dataframe().apply(boxcox_normmax, method=optim_method)
+        elif isinstance(lmbda, Sequence):
+            raise_if(len(lmbda) != len(data.pd_dataframe().columns),
+                     "lmbda should have one value per dimension (ie. column or variable) of the time series",
+                     logger)
+        else:
+            # Replicate lmbda to match dimensions of the time series
+            lmbda = [lmbda] * len(data.pd_dataframe().columns)
+
+        self._lmbda = lmbda
+
+        return self
+
+    def transform(self, data: TimeSeries, *args, **kwargs) -> TimeSeries:
+        super().transform(data, *args, **kwargs)
+
+        def _boxcox_wrapper(col):
+            idx = data.pd_dataframe().columns.get_loc(col.name)  # get index from col name
+            return boxcox(col, self._lmbda[idx])
+
+        return TimeSeries.from_dataframe(data.pd_dataframe().apply(_boxcox_wrapper))
+
+    def inverse_transform(self, data: TimeSeries, *args, **kwargs) -> TimeSeries:
+        super().inverse_transform(data, *args, *kwargs)
+
+        def _inv_boxcox_wrapper(col):
+            idx = data.pd_dataframe().columns.get_loc(col.name)  # get index from col name
```

Comment:
beyond the fact that it's currently copying the full DF for each column, isnt' there a cleaner way than using the name to get the column index? I'm sure there is a way to iterate over the column indices directly.

Reply:
> isnt' there a cleaner way than using the name to get the column index?

Good question, from within `apply()` I don't think so.

It's possible to iterate over the columns with a for loop instead and select the corresponding column (and perform a map on it). Something like:

```
for index in range(0, data.width):
   columnSeries = data._df.iloc[: , index]
   columnSeries.map(…)
```

Two concerns regarding this solution though:
- This could be less efficient than `apply()` for time series with many dimensions (though it depends on the implementation of `apply()`), since it iterates on the columns sequentially when there really isn't the need to enforce that.
- I'll have to check, but I think calling `iloc` this way returns a view, and calling `pandas.Series.map()` might actually modify the dataframe. So we'd still need to make at least one copy of the dataframe.

There are two other solutions I could think of:
1. Have lmbda be a `dict` that would map column names to lambda value. But some transforms loose the name of the column, so this is likely to fail when using `BoxCox` as part of a `Pipeline`.
2. Rename columns with their index instead of whatever name they have (can be done in place without copying anything), use `apply()` and then revert to the original names.

I don't think that 1. is very good, but 2. might actually be better than using the column name to compute an index. WDYT ?

Reply:
OK fair enough, then I would even say we could leave it using your current approach (it doesn't look super clean, but neither does 1. or 2.). I would still use `df_` in most places instead of `pd_dataframe()`, because you know what you're doing here so there's little risk and it can have a big performance impact.